sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -3805,3 +3799,233 @@ object ArrayUnion {
     new GenericArrayData(arrayBuffer)
   }
 }
+
+/**
+ * Returns an array of the elements in the intersect of x and y, without duplicates
+ */
+@ExpressionDescription(
+  usage = """
+  _FUNC_(array1, array2) - Returns an array of the elements in array1 but not in array2,
+    without duplicates.
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 2, 3), array(1, 3, 5));
+       array(2)
+  """,
+  since = "2.4.0")
+case class ArrayExcept(left: Expression, right: Expression) extends ArraySetLike {
+  override def dataType: DataType = left.dataType
+
+  var hsInt: OpenHashSet[Int] = _
+  var hsLong: OpenHashSet[Long] = _
+
+  def assignInt(array: ArrayData, idx: Int, resultArray: ArrayData, pos: Int): Boolean = {
+    val elem = array.getInt(idx)
+    if (!hsInt.contains(elem)) {
+      if (resultArray != null) {
+        resultArray.setInt(pos, elem)
+      }
+      hsInt.add(elem)
+      true
+    } else {
+      false
+    }
+  }
+
+  def assignLong(array: ArrayData, idx: Int, resultArray: ArrayData, pos: Int): Boolean = {
+    val elem = array.getLong(idx)
+    if (!hsLong.contains(elem)) {
+      if (resultArray != null) {
+        resultArray.setLong(pos, elem)
+      }
+      hsLong.add(elem)
+      true
+    } else {
+      false
+    }
+  }
+
+  def evalIntLongPrimitiveType(
+      array1: ArrayData,
+      array2: ArrayData,
+      resultArray: ArrayData,
+      isLongType: Boolean): Int = {
+    // store elements into resultArray
+    var notFoundNullElement = true
+    var i = 0
+    while (i < array2.numElements()) {
+      if (array2.isNullAt(i)) {
+        notFoundNullElement = false
+      } else {
+        val assigned = if (!isLongType) {
+          hsInt.add(array2.getInt(i))
+        } else {
+          hsLong.add(array2.getLong(i))
+        }
+      }
+      i += 1
+    }
+    var pos = 0
+    i = 0
+    while (i < array1.numElements()) {
+      if (array1.isNullAt(i)) {
+        if (notFoundNullElement) {
+          if (resultArray != null) {
+            resultArray.setNullAt(pos)
+          }
+          pos += 1
+          notFoundNullElement = false
+        }
+      } else {
+        val assigned = if (!isLongType) {
+          assignInt(array1, i, resultArray, pos)
+        } else {
+          assignLong(array1, i, resultArray, pos)
+        }
+        if (assigned) {
+          pos += 1
+        }
+      }
+      i += 1
+    }
+    pos
+  }
+
+  @transient lazy val evalExcept: (ArrayData, ArrayData) => ArrayData = {
+    if (elementTypeSupportEquals) {
+      elementType match {
+        case IntegerType =>
```

Comment:
why we only optimize for int and long? how about byte, short, float, double?

Reply:
if they all worth to optimize, I feel we should not optimize it in the interpreted code path and leave it to codegen path. That's a strong reason to add codegen support.

Reply:
Yeah, since the generated code also uses this path, we can optimize byte/short/boolean. However, it is not easy to optimize float and double with `OpenHashSet`. Let me look for hashset than can store primitive float/double.

Reply:
Ah, if we usually use the generated code, we will optimize them only in the generated code. In the interpreted path, we leave only the generic case as a backup.

Reply:
Let me try support float/double in `OpenHashSet`.